File: py/classical/double_transposition.py
from collections import Counter


def _parse_key(key: str) -> list[int]:
    """
    Convert a keyword or numeric list to a 0-indexed reading order.
    Keyword: sorted by alphabetical rank — e.g. "CAB" → [1, 2, 0]
    Numeric: "3,1,2" → [2, 0, 1] (0-indexed)
    """
    key = key.strip()
    if "," in key:
        parts = [int(x.strip()) for x in key.split(",")]
        order = [p - 1 for p in parts]
        seen: set[int] = set()
        for v in order:
            if v < 0 or v >= len(order):
                raise ValueError(f"Index {v + 1} out of range (key length {len(order)})")
            if v in seen:
                raise ValueError(f"Duplicate index {v + 1}")
            seen.add(v)
        return order
    else:
        letters = list(key.upper())
        return sorted(range(len(letters)), key=lambda i: letters[i])


def _row_col_encrypt(text: str, row_order: list[int], col_order: list[int]) -> tuple[str, str]:
    """
    Fill grid row-by-row, permute rows then columns, read row-by-row.
    Returns (after_row_perm, ciphertext).
    """
    n_rows = len(row_order)
    n_cols = len(col_order)

    grid = [list(text[r * n_cols:(r + 1) * n_cols]) for r in range(n_rows)]

    # Step 1: permute rows
    row_permed = [grid[row_order[i]] for i in range(n_rows)]

    after_row_perm = "".join("".join(row) for row in row_permed)

    # Step 2: permute columns
    cipher = "".join(row_permed[r][c] for r in range(n_rows) for c in col_order)

    return after_row_perm, cipher


def _row_col_decrypt(text: str, row_order: list[int], col_order: list[int]) -> tuple[str, str]:
    """
    Reverse _row_col_encrypt using inverse permutations.
    Returns (after_col_restore, plaintext).
    """
    n_rows = len(row_order)
    n_cols = len(col_order)

    c_grid = [list(text[r * n_cols:(r + 1) * n_cols]) for r in range(n_rows)]

    # Inverse column permutation: restore rowPermed state
    inv_col_grid = [[""] * n_cols for _ in range(n_rows)]
    for r in range(n_rows):
        for j, c in enumerate(col_order):
            inv_col_grid[r][c] = c_grid[r][j]

    after_col_restore = "".join("".join(row) for row in inv_col_grid)

    # Inverse row permutation: row_order[i]=r means inv_row[r]=i
    inv_row = [0] * n_rows
    for i, r in enumerate(row_order):
        inv_row[r] = i

    plain = "".join("".join(inv_col_grid[inv_row[r]]) for r in range(n_rows))
    return after_col_restore, plain


def handle(operation: str, params: dict) -> dict:
    op = operation.lower()

    if op == "encrypt":
        # Replace spaces with X so word boundaries survive encryption.
        plaintext    = params.get("plaintext", "").upper()
        plaintext    = plaintext.strip().replace(" ", "X")
        original_len = len(plaintext)
        key1         = params.get("key1", "")
        key2         = params.get("key2", "")

        row_order = _parse_key(key1)
        col_order = _parse_key(key2)

        n_rows = len(row_order)
        n_cols = len(col_order)
        total  = n_rows * n_cols

        if original_len > total:
            raise ValueError(
                f"Plaintext ({original_len} chars) exceeds grid capacity "
                f"{n_rows}×{n_cols}={total}; use longer keys."
            )

        # Pad to fill grid exactly.
        plaintext = plaintext.ljust(total, "X")

        after_row_perm, cipher = _row_col_encrypt(plaintext, row_order, col_order)

        return {
            "after_row_permutation": after_row_perm,
            "ciphertext":            cipher,
            "key1_order":            str([i + 1 for i in row_order]),
            "key2_order":            str([i + 1 for i in col_order]),
            "original_length":       str(original_len),
            "grid_size":             f"{n_rows} rows × {n_cols} cols",
        }

    elif op == "decrypt":
        ciphertext = params.get("ciphertext", "").upper()
        key1       = params.get("key1", "")
        key2       = params.get("key2", "")

        row_order = _parse_key(key1)
        col_order = _parse_key(key2)

        n_rows = len(row_order)
        n_cols = len(col_order)
        expected = n_rows * n_cols
        if len(ciphertext) != expected:
            raise ValueError(
                f"Ciphertext length {len(ciphertext)} does not match grid "
                f"{n_rows}×{n_cols}={expected}"
            )

        after_col_restore, plain = _row_col_decrypt(ciphertext, row_order, col_order)

        # Exact-length trim preferred to avoid eating real trailing X's.
        orig_len_str = params.get("original_length", "").strip()
        if orig_len_str:
            try:
                orig_len = int(orig_len_str)
                if 0 <= orig_len <= len(plain):
                    stripped     = plain[:orig_len]
                    padding_note = (f"exact trim to original length {orig_len} "
                                    f"(removed {len(plain) - orig_len} pad chars)")
                else:
                    stripped     = plain
                    padding_note = "original_length out of range; no trimming applied"
            except ValueError:
                stripped     = plain
                padding_note = "original_length invalid; no trimming applied"
        else:
            stripped     = plain.rstrip("X")
            padding_note = (f"heuristic trim (no original_length supplied): "
                            f"removed {len(plain) - len(stripped)} trailing X chars")

        return {
            "after_col_restore": after_col_restore,
            "plaintext":         stripped,
            "padding_note":      padding_note,
        }

    elif op == "frequency":
        ciphertext = params.get("ciphertext", "").upper()
        counts     = Counter(ch for ch in ciphertext if ch.isalpha())
        total      = sum(counts.values())
        if total == 0:
            raise ValueError("No alphabetic characters in ciphertext.")

        # English letter frequency order and reference percentages.
        english_order = list("ETAOINSHRDLCUMWFGYPBVKJXQZ")
        english_freq  = {
            'E': 12.70, 'T': 9.06, 'A': 8.17, 'O': 7.51, 'I': 6.97,
            'N': 6.75,  'S': 6.33, 'H': 6.09, 'R': 5.99, 'D': 4.25,
            'L': 4.03,  'C': 2.78, 'U': 2.76, 'M': 2.41, 'W': 2.36,
            'F': 2.23,  'G': 2.02, 'Y': 1.97, 'P': 1.93, 'B': 1.49,
            'V': 0.98,  'K': 0.77, 'J': 0.15, 'X': 0.15, 'Q': 0.10,
            'Z': 0.07,
        }

        ranked = counts.most_common()

        # Frequency table with English comparison.
        table_lines = ["Letter  Count   Cipher%   English%",
                       "------  -----  --------  ---------"]
        for ch, cnt in ranked:
            pct = cnt / total * 100
            eng = english_freq.get(ch, 0.0)
            table_lines.append(f"  {ch}     {cnt:4d}    {pct:5.1f}%    {eng:5.2f}%")

        # Index of Coincidence.
        ioc = (sum(c * (c - 1) for c in counts.values()) /
               (total * (total - 1))) if total > 1 else 0.0
        if ioc >= 0.060:
            ioc_note = " (≈ English — likely transposition or mono-alphabetic substitution)"
        elif ioc >= 0.045:
            ioc_note = " (moderate — possibly short Vigenère or polyalphabetic)"
        else:
            ioc_note = " (low — likely polyalphabetic / random)"

        # Frequency-mapping guess: map cipher's top-N to English's top-N.
        mapping = {ch: english_order[i] for i, (ch, _) in enumerate(ranked)
                   if i < len(english_order)}
        guessed = "".join(mapping.get(ch, ch) for ch in ciphertext if ch.isalpha())

        # Mapping legend.
        legend_lines = ["Cipher → Guess (by frequency rank)"]
        for i, (ch, _) in enumerate(ranked):
            if i >= len(english_order):
                break
            legend_lines.append(f"  {ch} → {english_order[i]}")

        return {
            "frequency_table":   "\n".join(table_lines),
            "ioc":               f"{ioc:.4f}{ioc_note}",
            "mapping_legend":    "\n".join(legend_lines),
            "guessed_plaintext": guessed,
            "note": (
                "For transposition ciphers letter frequencies are preserved from the "
                "plaintext. A high IoC (≥0.060) confirms transposition. The guessed "
                "plaintext applies a simple frequency substitution — it is useful mainly "
                "for mono-alphabetic substitution ciphers, not pure transposition."
            ),
        }

    else:
        raise ValueError(f"Unknown operation '{operation}' for double transposition.")

File: internal/engine/classical/double_transposition.go
package classical

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
)

func DoubleTransposition(operation string, params map[string]string) (map[string]string, error) {
	op := strings.ToLower(operation)

	switch op {
	case "encrypt":
		// Replace spaces with X (spaces become visible in decrypted output as X).
		plaintext := strings.ToUpper(strings.ReplaceAll(strings.TrimSpace(params["plaintext"]), " ", "X"))
		originalLen := len(plaintext) // before padding

		rowOrder, err := parseKey(params["key1"])
		if err != nil {
			return nil, fmt.Errorf("key1: %w", err)
		}
		colOrder, err := parseKey(params["key2"])
		if err != nil {
			return nil, fmt.Errorf("key2: %w", err)
		}

		nRows := len(rowOrder)
		nCols := len(colOrder)
		total := nRows * nCols

		if originalLen > total {
			return nil, fmt.Errorf(
				"plaintext (%d chars) exceeds grid capacity %d×%d=%d; use longer keys",
				originalLen, nRows, nCols, total)
		}

		// Pad to fill grid exactly.
		if originalLen < total {
			plaintext += strings.Repeat("X", total-originalLen)
		}

		afterRowPerm, cipher := rowColEncrypt(plaintext, rowOrder, colOrder)

		return map[string]string{
			"after_row_permutation": afterRowPerm,
			"ciphertext":            cipher,
			"key1_order":            formatOrder(rowOrder),
			"key2_order":            formatOrder(colOrder),
			"original_length":       strconv.Itoa(originalLen),
			"grid_size":             fmt.Sprintf("%d rows × %d cols", nRows, nCols),
		}, nil

	case "decrypt":
		ciphertext := strings.ToUpper(params["ciphertext"])
		rowOrder, err := parseKey(params["key1"])
		if err != nil {
			return nil, fmt.Errorf("key1: %w", err)
		}
		colOrder, err := parseKey(params["key2"])
		if err != nil {
			return nil, fmt.Errorf("key2: %w", err)
		}

		nRows := len(rowOrder)
		nCols := len(colOrder)
		if len(ciphertext) != nRows*nCols {
			return nil, fmt.Errorf(
				"ciphertext length %d does not match grid %d×%d=%d",
				len(ciphertext), nRows, nCols, nRows*nCols)
		}

		afterColRestore, plain := rowColDecrypt(ciphertext, rowOrder, colOrder)

		// Trim padding: exact slice preferred to avoid eating real trailing X's.
		var stripped, paddingNote string
		if s := strings.TrimSpace(params["original_length"]); s != "" {
			origLen, err := strconv.Atoi(s)
			if err == nil && origLen >= 0 && origLen <= len(plain) {
				stripped = plain[:origLen]
				paddingNote = fmt.Sprintf("exact trim to original length %d (removed %d pad chars)", origLen, len(plain)-origLen)
			} else {
				stripped = plain
				paddingNote = "original_length invalid or out of range; no trimming applied"
			}
		} else {
			stripped = strings.TrimRight(plain, "X")
			paddingNote = fmt.Sprintf("heuristic trim (no original_length supplied): removed %d trailing X chars", len(plain)-len(stripped))
		}

		return map[string]string{
			"after_col_restore": afterColRestore,
			"plaintext":         stripped,
			"padding_note":      paddingNote,
		}, nil

	case "frequency":
		ciphertext := strings.ToUpper(params["ciphertext"])
		counts := make(map[rune]int)
		total := 0
		for _, ch := range ciphertext {
			if ch >= 'A' && ch <= 'Z' {
				counts[ch]++
				total++
			}
		}
		if total == 0 {
			return nil, fmt.Errorf("no alphabetic characters in ciphertext")
		}

		// English letter frequency order (most→least common).
		englishOrder := []rune("ETAOINSHRDLCUMWFGYPBVKJXQZ")

		type pair struct {
			ch  rune
			cnt int
		}
		var pairs []pair
		for ch, cnt := range counts {
			pairs = append(pairs, pair{ch, cnt})
		}
		sort.Slice(pairs, func(i, j int) bool { return pairs[i].cnt > pairs[j].cnt })

		// English reference frequencies (%).
		englishFreq := map[rune]float64{
			'E': 12.70, 'T': 9.06, 'A': 8.17, 'O': 7.51, 'I': 6.97,
			'N': 6.75, 'S': 6.33, 'H': 6.09, 'R': 5.99, 'D': 4.25,
			'L': 4.03, 'C': 2.78, 'U': 2.76, 'M': 2.41, 'W': 2.36,
			'F': 2.23, 'G': 2.02, 'Y': 1.97, 'P': 1.93, 'B': 1.49,
			'V': 0.98, 'K': 0.77, 'J': 0.15, 'X': 0.15, 'Q': 0.10,
			'Z': 0.07,
		}

		// Frequency table with English comparison.
		var tableLines strings.Builder
		tableLines.WriteString("Letter  Count   Cipher%   English%\n")
		tableLines.WriteString("------  -----  --------  ---------\n")
		for _, p := range pairs {
			pct := float64(p.cnt) / float64(total) * 100
			eng := englishFreq[p.ch]
			tableLines.WriteString(fmt.Sprintf("  %c     %4d    %5.1f%%    %5.2f%%\n", p.ch, p.cnt, pct, eng))
		}

		// Index of Coincidence (IoC). English prose ≈ 0.065; random ≈ 0.038.
		ioc := 0.0
		if total > 1 {
			for _, cnt := range counts {
				ioc += float64(cnt) * float64(cnt-1)
			}
			ioc /= float64(total) * float64(total-1)
		}
		ioStr := fmt.Sprintf("%.4f", ioc)
		ioNote := ""
		switch {
		case ioc >= 0.060:
			ioNote = " (≈ English — likely transposition or mono-alphabetic substitution)"
		case ioc >= 0.045:
			ioNote = " (moderate — possibly short Vigenère or polyalphabetic)"
		default:
			ioNote = " (low — likely polyalphabetic / random)"
		}

		// Frequency-mapping guess: map ciphertext's top-N letters to English's top-N.
		// For a transposition cipher this is a no-op (frequencies are preserved),
		// but it gives a rough plaintext guess for substitution-like analysis.
		mapping := make(map[rune]rune)
		for i, p := range pairs {
			if i >= len(englishOrder) {
				break
			}
			mapping[p.ch] = englishOrder[i]
		}
		var guessed strings.Builder
		for _, ch := range ciphertext {
			if ch >= 'A' && ch <= 'Z' {
				if mapped, ok := mapping[ch]; ok {
					guessed.WriteRune(mapped)
				} else {
					guessed.WriteRune(ch)
				}
			}
		}

		// Build mapping legend.
		var legendLines strings.Builder
		legendLines.WriteString("Cipher → Guess (by frequency rank)\n")
		for i, p := range pairs {
			if i >= len(englishOrder) {
				break
			}
			legendLines.WriteString(fmt.Sprintf("  %c → %c\n", p.ch, englishOrder[i]))
		}

		return map[string]string{
			"frequency_table":   strings.TrimRight(tableLines.String(), "\n"),
			"ioc":               ioStr + ioNote,
			"mapping_legend":    strings.TrimRight(legendLines.String(), "\n"),
			"guessed_plaintext": guessed.String(),
			"note": "For transposition ciphers letter frequencies are preserved from the " +
				"plaintext. A high IoC (≥0.060) confirms transposition. The guessed plaintext " +
				"applies a simple frequency substitution — it is useful mainly for mono-alphabetic " +
				"substitution ciphers, not pure transposition.",
		}, nil

	default:
		return nil, fmt.Errorf("unknown operation %q for double transposition", operation)
	}
}

// rowColEncrypt fills a grid row-by-row, permutes rows then columns, reads row-by-row.
// Returns (afterRowPerm, ciphertext).
func rowColEncrypt(text string, rowOrder, colOrder []int) (string, string) {
	nRows := len(rowOrder)
	nCols := len(colOrder)
	runes := []rune(text)

	// Original grid
	grid := make([][]rune, nRows)
	for r := 0; r < nRows; r++ {
		grid[r] = runes[r*nCols : (r+1)*nCols]
	}

	// Step 1: permute rows — new_grid[i] = grid[rowOrder[i]]
	rowPermed := make([][]rune, nRows)
	for i, r := range rowOrder {
		rowPermed[i] = grid[r]
	}

	// Intermediate: read row-permed grid row by row
	var sbMid strings.Builder
	for _, row := range rowPermed {
		sbMid.WriteString(string(row))
	}

	// Step 2: permute columns — cipher[r][j] = rowPermed[r][colOrder[j]]
	var sbOut strings.Builder
	for r := 0; r < nRows; r++ {
		for _, c := range colOrder {
			sbOut.WriteRune(rowPermed[r][c])
		}
	}

	return sbMid.String(), sbOut.String()
}

// rowColDecrypt reverses rowColEncrypt using inverse permutations.
// Returns (afterColRestore, plaintext).
func rowColDecrypt(text string, rowOrder, colOrder []int) (string, string) {
	nRows := len(rowOrder)
	nCols := len(colOrder)
	runes := []rune(text)

	// Fill cipher grid row by row
	cGrid := make([][]rune, nRows)
	for r := 0; r < nRows; r++ {
		cGrid[r] = runes[r*nCols : (r+1)*nCols]
	}

	// Inverse column permutation: colOrder[j]=c means col j of result came from col c of input.
	// Restore: rowPermed[r][colOrder[j]] = cGrid[r][j]
	invCol := make([][]rune, nRows)
	for r := 0; r < nRows; r++ {
		invCol[r] = make([]rune, nCols)
		for j, c := range colOrder {
			invCol[r][c] = cGrid[r][j]
		}
	}

	// Intermediate: after restoring columns (= after-row-perm state)
	var sbMid strings.Builder
	for _, row := range invCol {
		sbMid.WriteString(string(row))
	}

	// Inverse row permutation: rowOrder[i]=r means row i of rowPermed came from row r of original.
	// Restore: grid[rowOrder[i]] = rowPermed[i] → grid[r] = invCol[invRow[r]]
	invRow := make([]int, nRows)
	for i, r := range rowOrder {
		invRow[r] = i
	}

	var sbOut strings.Builder
	for r := 0; r < nRows; r++ {
		sbOut.WriteString(string(invCol[invRow[r]]))
	}

	return sbMid.String(), sbOut.String()
}

// parseKey converts a keyword (e.g. "KEY") or numeric list (e.g. "3,1,2") to 0-indexed reading order.
// For a keyword the result is indices sorted by alphabetical rank of each letter.
// E.g. "CAB" → C=rank2, A=rank0, B=rank1 → reading order [1, 2, 0].
func parseKey(key string) ([]int, error) {
	key = strings.TrimSpace(key)
	if strings.Contains(key, ",") {
		parts := strings.Split(key, ",")
		order := make([]int, len(parts))
		for i, p := range parts {
			n, err := strconv.Atoi(strings.TrimSpace(p))
			if err != nil {
				return nil, fmt.Errorf("invalid numeric key segment %q", p)
			}
			order[i] = n - 1
		}
		seen := make(map[int]bool)
		for _, v := range order {
			if v < 0 || v >= len(order) {
				return nil, fmt.Errorf("index %d out of range (key length %d)", v+1, len(order))
			}
			if seen[v] {
				return nil, fmt.Errorf("duplicate index %d", v+1)
			}
			seen[v] = true
		}
		return order, nil
	}

	letters := []rune(strings.ToUpper(key))
	n := len(letters)
	temp := make([]rune, n)
	copy(temp, letters)
	sort.Slice(temp, func(i, j int) bool { return temp[i] < temp[j] })

	used := make([]bool, n)
	rankOf := make([]int, n)
	for i, ch := range temp {
		for j, orig := range letters {
			if orig == ch && !used[j] {
				rankOf[j] = i
				used[j] = true
				break
			}
		}
	}
	indices := make([]int, n)
	for i := range indices {
		indices[i] = i
	}
	sort.Slice(indices, func(i, j int) bool { return rankOf[indices[i]] < rankOf[indices[j]] })
	return indices, nil
}

func formatOrder(order []int) string {
	parts := make([]string, len(order))
	for i, v := range order {
		parts[i] = strconv.Itoa(v + 1)
	}
	return "[" + strings.Join(parts, " ") + "]"
}
